src/platforms/macos/daemon/dnsutilsmacos.cpp

```diff
@@ -19,205 +17,52 @@ Logger logger("DnsUtilsMacos");
 }
 
-static void cfDictSetStringList(CFMutableDictionaryRef dict, CFStringRef name,
-                                const QStringList& valueList) {
-  if (valueList.isEmpty()) {
-    return;
-  }
-
-  CFMutableArrayRef array;
-  array = CFArrayCreateMutable(kCFAllocatorSystemDefault, 0,
-                               &kCFTypeArrayCallBacks);
-  if (array == nullptr) {
-    return;
-  }
-
-  for (const QString& rstring : valueList) {
-    CFStringRef cfAddr = CFStringCreateWithCString(kCFAllocatorSystemDefault,
-                                                   qUtf8Printable(rstring),
-                                                   kCFStringEncodingUTF8);
-    CFArrayAppendValue(array, cfAddr);
-    CFRelease(cfAddr);
+void DnsUtilsMacos::dnsManagerStdoutReady() {
+  for (;;) {
+    QByteArray line = m_dnsManager.readLine();
+    if (line.length() <= 0) {
+      break;
+    }
+    logger.debug() << QString::fromUtf8(line);
   }
-  CFDictionarySetValue(dict, name, array);
-  CFRelease(array);
 }
 
 bool DnsUtilsMacos::updateResolvers(const QString& ifname,
                                     const QList<QHostAddress>& resolvers) {
   Q_UNUSED(ifname);
 
-  // Get the list of current network services.
-  CFArrayRef netServices = SCDynamicStoreCopyKeyList(
-      m_scStore, CFSTR("Setup:/Network/Service/[0-9A-F-]+"));
-  if (netServices == nullptr) {
-    return false;
-  }
-  auto serviceGuard = qScopeGuard([&] { CFRelease(netServices); });
-
-  // Prepare the DNS configuration.
-  CFMutableDictionaryRef dnsConfig = CFDictionaryCreateMutable(
-      kCFAllocatorSystemDefault, 0, &kCFCopyStringDictionaryKeyCallBacks,
-      &kCFTypeDictionaryValueCallBacks);
-  auto configGuard = qScopeGuard([&] { CFRelease(dnsConfig); });
-  QStringList list;
+  // Prepare the process arguments.
+  QStringList args;
+  args.append("macosdnsmanager");
   for (const QHostAddress& addr : resolvers) {
-    list.append(addr.toString());
+    args.append(addr.toString());
   }
-  cfDictSetStringList(dnsConfig, kSCPropNetDNSServerAddresses, list);
-  cfDictSetString(dnsConfig, kSCPropNetDNSDomainName, "lan");
 
-  // Backup each network service's DNS config, and replace it with ours.
-  for (CFIndex i = 0; i < CFArrayGetCount(netServices); i++) {
-    QString service = cfParseString(CFArrayGetValueAtIndex(netServices, i));
-    QString uuid = service.section('/', 3, 3);
-    if (uuid.isEmpty()) {
-      continue;
-    }
-    backupService(uuid);
-
-    logger.debug() << "Setting DNS config for" << uuid;
-    CFStringRef dnsPath = CFStringCreateWithFormat(
-        kCFAllocatorSystemDefault, nullptr,
-        CFSTR("Setup:/Network/Service/%s/DNS"), qPrintable(uuid));
-    if (!dnsPath) {
-      continue;
-    }
-    SCDynamicStoreSetValue(m_scStore, dnsPath, dnsConfig);
-    CFRelease(dnsPath);
-  }
-  return true;
+  // Launch the DNS manager process.
+  m_dnsManager.start(qApp->applicationFilePath(), args);
+  return m_dnsManager.waitForStarted();
```

Comment:
Is waitForStarted's default timeout OK?

Reply:
I believe so, normally it should only take a few milliseconds to launch a new process, so if we encounter the 30s default timeout here something has gone horribly wrong.

Reply:
A timeout here may need be handled by requesting process termination of the child process, so it can cleanup.

Reply:
Done!